### PR TITLE
fix(TFD-2130) : Wrap schema registry factory in a ThreadLocal

### PR DESCRIPTION
**Pull request by lbourgeois**
**What is the current behavior?** (You can also link to an open issue here)
- schema registry in lazy avro coder is stored in a static variable
- it may cause problems if multiple compilations occurs at the same time

**What is the new behavior?**
- fix these potential problems by wrapping factory in ThreadLocal

**Please check if the PR fulfills these requirements**

- [X] The commit message follows Talend standard
- [X] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features) ?
- [ ] The code coverage on new code >75%
- [ ] The new code does not introduce new technical issues (sonar / eslint)

**What kind of change does this PR introduce?**

- [X] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build / CI related changes
- [ ] Other... Please describe:

**Does this PR introduce a breaking change?**

- [ ] Yes
- [ ] No

If this PR contains a breaking change, please describe the impact and migration path for existing applications: ...

**Other information**:

### CODE REVIEW

File: core/components-adapter-beam/src/main/java/org/talend/components/adapter/beam/coders/LazyAvroCoder.java [hunk 1]
```diff
@@ -58,7 +58,7 @@
 
     private static final Logger LOG = LoggerFactory.getLogger(LazyAvroCoder.class);
 
-    private transient static Supplier<AvroSchemaHolder> schemaRegistryFactory;
+    private transient static ThreadLocal<Supplier<AvroSchemaHolder>> threadSchemaRegistryFactory;
```
Comment (RyanSkraba) [inline core/components-adapter-beam/src/main/java/org/talend/components/adapter/beam/coders/LazyAvroCoder.java]: Rename registry! `threadSchemaSupplier`?

File: core/components-adapter-beam/src/main/java/org/talend/components/adapter/beam/coders/LazyAvroCoder.java [hunk 2]
```diff
@@ -71,24 +71,29 @@ protected LazyAvroCoder() {
...
+        threadSchemaRegistryFactory.remove();
     }
 
     public static void setSchemaRegistry(Supplier<AvroSchemaHolder> factory) {
         // Ensure that the schema registry has not already been set.
-        if (LazyAvroCoder.schemaRegistryFactory != null)
-            throw BeamAdapterErrorCode.createSchemaRegistryAlreadyExists(null, schemaRegistryFactory.toString());
-        schemaRegistryFactory = factory;
+        if (LazyAvroCoder.threadSchemaRegistryFactory != null)
+            throw BeamAdapterErrorCode.createSchemaRegistryAlreadyExists(null, threadSchemaRegistryFactory.get().toString());
```
Comment (RyanSkraba) [inline core/components-adapter-beam/src/main/java/org/talend/components/adapter/beam/coders/LazyAvroCoder.java]: BeamAdapterErrorCode.setSchemaSupplierWhenAlreadyExists

File: core/components-adapter-beam/src/main/java/org/talend/components/adapter/beam/coders/LazyAvroCoder.java [hunk 3]
```diff
@@ -71,24 +71,29 @@ protected LazyAvroCoder() {
     }
 
     public static void resetSchemaRegistry() {
```
Comment (RyanSkraba) [inline core/components-adapter-beam/src/main/java/org/talend/components/adapter/beam/coders/LazyAvroCoder.java]: resetSchemaSupplier

File: core/components-adapter-beam/src/main/java/org/talend/components/adapter/beam/coders/LazyAvroCoder.java [hunk 4]
```diff
@@ -71,24 +71,29 @@ protected LazyAvroCoder() {
...
+        Supplier<AvroSchemaHolder> schemaRegistryFactory = threadSchemaRegistryFactory.get();
         if (schemaRegistryFactory instanceof StaticSchemaHolderSupplier) {
             ((StaticSchemaHolderSupplier) schemaRegistryFactory).reset();
         }
-        schemaRegistryFactory = null;
+        threadSchemaRegistryFactory.remove();
     }
 
     public static void setSchemaRegistry(Supplier<AvroSchemaHolder> factory) {
```
Comment (RyanSkraba) [inline core/components-adapter-beam/src/main/java/org/talend/components/adapter/beam/coders/LazyAvroCoder.java]: setSchemaSupplier

File: core/components-adapter-beam/src/main/java/org/talend/components/adapter/beam/BeamAdapterErrorCode.java [hunk 5]
```diff
@@ -77,7 +77,7 @@ public String getCode() {
      * @param existing The current SchemaRegistrySupplier.
      * @return An exception corresponding to the error code.
      */
-    public static TalendRuntimeException createSchemaRegistryAlreadyExists(Throwable cause, String existing) {
+    public static TalendRuntimeException createSchemaSupplierAlreadyExists(Throwable cause, String existing) {
         return new TalendMsgRuntimeException(cause, SCHEMA_REGISTRY_ALREADY_EXISTS,
```
Comment (RyanSkraba) [inline core/components-adapter-beam/src/main/java/org/talend/components/adapter/beam/BeamAdapterErrorCode.java]: More Registry!  See JavaDoc and error msg too.

**Comment by build-talend-common** [comment]
Pull request test result: [failure](https://ci-common.datapwn.com/job/components_pull-request_master/1196/) [:x:](https://ci-common.datapwn.com/job/components_pull-request_master/1196/)

**Comment by build-talend-common** [comment]
Pull request test result: [success](https://ci-common.datapwn.com/job/components_pull-request_master/1195/) [:star:](https://ci-common.datapwn.com/job/components_pull-request_master/1195/) 
[Sonar results](https://sonar.datapwn.com/overview?id=org.talend.components%3Acomponents%3Afix%2FTFD2130_thread_local_schema_registry)

**Comment by build-talend-common** [comment]
Pull request test result: [failure](https://ci-common.datapwn.com/job/components_pull-request_master/1197/) [:x:](https://ci-common.datapwn.com/job/components_pull-request_master/1197/)

**Comment by build-talend-common** [comment]
Pull request test result: [success](https://ci-common.datapwn.com/job/components_pull-request_master/1198/) [:star:](https://ci-common.datapwn.com/job/components_pull-request_master/1198/) 
[Sonar results](https://sonar.datapwn.com/overview?id=org.talend.components%3Acomponents%3Afix%2FTFD2130_thread_local_schema_registry)

**Pull request merged**